spec/OpenLineage.yml

```diff
@@ -0,0 +1,250 @@
+openapi: 3.0.2
+info:
+  title: OpenLineage
+  version: 0.0.1
+  description: OpenLineage is an open source **lineage and metadata collection API** for the data ecosystem.
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+paths:
+  /events:
+    post:
+      summary: Updates the state of a run
+      description: Updates a run state for a job.
+      operationId: postRunStateUpdate
+      tags:
+        - OpenLineage
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/RunStateUpdates'
+      responses:
+        '200':
+          description: OK
+
+components:
+  schemas:
+    RunStateUpdates:
+      type: array
+      items:
+        $ref: '#/components/schemas/RunStateUpdate'
+
+    RunStateUpdate:
+      type: object
+      properties:
+        runId:
+          description: the unique identifier for that run
+          type: string
+          format: uuid
+        jobId:
+          $ref: '#/components/schemas/JobId'
+        transition:
+          description: transitions the run state
+          type: string #TODO: add enum
+          example: 'START|RUNNING|COMPLETE|ABORT|FAIL'
+        transitionTime:
+          type: string
+          format: date-time
+        origin:
+          $ref: '#/components/schemas/Origin'
+        job:
+          $ref: '#/components/schemas/Job'
+        inputs:
+          description: The set of **input** datasets.
+          type: array
+          items:
+            $ref: '#/components/schemas/Dataset'
+        outputs:
+          description: The set of **output** datasets.
+          type: array
+          items:
+            $ref: '#/components/schemas/Dataset'
+        facets:
+          description: The run facets.
+          type: object
+          properties:
+            nominalTime:
+              $ref: '#/components/schemas/NominalTimeRunFacet'
+            parent:
+              $ref: '#/components/schemas/ParentRunFacet'
+          additionalProperties:
+            $ref: '#/components/schemas/CustomFacet'
+      required:
+        - runId
+        - jobId
+        - transition
+        - transition-time
+        - origin
+
+    Job:
+      type: object
+      properties:
+        description:
+          description: The description of the job.
+          type: string
+        facets:
+          description: The job facets.
+          type: object
+          properties:
+            sourceCodeLocation:
+              $ref: '#/components/schemas/SourceCodeLocationJobFacet'
+            sql:
+              $ref: '#/components/schemas/SQLJobFacet'
+          additionalProperties:
```

Comment:
This type of semi-structured data is really tough to model in strongly typed languages (e.g. graphql) since its tough to have arbitrary keys. How would you feel about facets having the following def. The downside is that we'd need to introduce some tooling but it'd make it facets more future proof.
```
    Job:
      type: object
      properties:
        description:
          description: The description of the job.
          type: string
        facets:
          description: The job facets.
          type: array
          items:
            $ref: '#/components/schemas/Facet'

    Facet:
      type: object
      properties:
        origin:
          description: The URI of the facet property
          type: string
          example: http://openlineage.com/facets/schema/postgres
        content:
          description: The content of the facet
          type: object
      required:
        - origin
```

Reply:
I'm drawing inspiration from https://github.com/graphql/graphql-spec/issues/635

Reply:
Heres an overview of some approaches to input union types: https://github.com/graphql/graphql-spec/blob/master/rfcs/InputUnion.md

Reply:
Yes, I agree. I am somewhat intending that in the definition of CustomFacet bellow with the origin and version fields that could actually be replaced by an URI to the definition.
It's akin to Map<String, CustomFacet> with CustomeFacet having some required fields.
The goal is to allow sending facets that are not part of the spec to strike a balance between standard facets that might be lossy and allowing adding all the details specific to a given system.